### ISSUE TITLE
How about using tools like jsLint to get better code quality?

### ISSUE THREAD

**Issue by eisbehr-** [node-red/node-red#1179]
While reviewing some parts of the code, I've noticed that there many different writings or maybe formatting misstakes in the code. Sometimes it's a real mess.

There are many different tools to get better code quality. How about to use one and create a fixed set or requirements once?

Some examples, found in just one file '_nodes/core/io/10-mqtt.js_': 

- Used different quotes:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L20-L21

- Sometimes spaces between operators, sometimes not, sometimes both:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L133-L135

- Using `===` or `==` ?
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L135

- Sometimes no space between _function_ and braces, sometimes spaces:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L150
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L166

- Sometimes spacec between parameters, sometimes don't:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L237
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L259

- Space between properties and value, or not:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L313-L314

- Or even more worse:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L337

- How to write an if?!
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L333
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L336

- Brace madness:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L376
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L382

- Unwanted spaces:
https://github.com/node-red/node-red/blob/master/nodes/core/io/10-mqtt.js#L386

And this is just what I saw in this one file. You can imagine how the whole code may look like. Tools like jsLint could help to improve the overall quality of node-red.

**Comment by knolleary** [comment]
Hi, I agree there may be occasional inconsistencies through the code, and this is something we've discussed amongst ourselves a number of times since the start of the project.

A lot of those issues are entirely cosmetic - such as the whitespace items you've highlighted - and have little to do with code quality.

I disagree with some of your criticisms - for example, what you call 'brace madness' can also be called 'being explicit so it is obvious what the logic is'. That is a lesson learnt from past experience.

Where there are functional consequences, we already use jshint in the build to catch those (such as improper use of `==`).

So, yes, introducing jslint is something we could consider. But please consider the tone of your issue report - it comes across as very negative criticism, rather than highlight the benefits of using such a tool.


Thanks!

**Comment by eisbehr-** [comment]
Of course, many things are more cosmetically in nature. But this is even a part of code quality. There are many projects using a detailed coding guides to keep the formatting clean. And this would totally make sense here too, if even one single fine has so many different writings in it.

Google made a pretty good job here IMO, for example. But there are others for sure, that could be adapted.
- https://google.github.io/styleguide/jsguide.html
- https://google.github.io/styleguide/javascriptguide.xml
- https://contribute.jquery.org/style-guide/js/
- https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style

I even know projects (_we even do it internally too_), that reject all changes with no proper formatting. Otherwise, it could be hard work to clean the code up later again.

And I meant "brackets" not "brace". My fault. But could you tell me the difference between those two, and why the first one is clearer to someone?

```JS
if ((msg.qos !== 0) && (msg.qos !== 1) && (msg.qos !== 2)) { }
if (msg.qos !== 0 && msg.qos !== 1 && msg.qos !== 2) { }

msg.retain = ((msg.retain === true) || (msg.retain === "true")) || false;
msg.retain = msg.retain === true || msg.retain === "true";
```

And sorry, I'm not a native speaker, so this may because you think it was a bit harsh.  
But it may not be free of a little criticism too. ;)

**Comment by knolleary** [comment]
I am not against the idea of linting; you don't need to explain the benefits.

You will find the core code of node-red is much more consistent - we do reject PRs that have bad code style; it just isn't a check we have fully automated. And some of the older code, or less-touched code may contain some inconsistencies we haven't tidied up. Do not assume this one file reflects the entire code base.

I'm not going to debate specific examples - I wanted to highlight that sometimes linting rules should not be enforced so tightly; most linting tools allow you to instrument the code with comments to bend the rules where it makes sense to.

I don't mind criticism, but I prefer constructive feedback.

**Comment by knolleary** [comment]
One final comment - if you check out our [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md) you'll see we prefer to discuss this sort of thing on the [mailing list](https://groups.google.com/forum/#!forum/node-red) or [slack team](https://nodered.org/slack). We try to keep the issue list for functional issues that need work.

Thanks!

**Comment by eisbehr-** [comment]
> rather than **highlight the benefits** of using such a tool.

...

> you **don't need to explain the benefits**.

What?!

And I would not confirm, that the of the code looks better. Lets make a quick scroll-to the main `red.js` file:
- https://github.com/node-red/node-red/blob/master/red.js#L18-L19
- https://github.com/node-red/node-red/blob/master/red.js#L22-L23 (against your own guide)
- https://github.com/node-red/node-red/blob/master/red.js#L116-L117
- https://github.com/node-red/node-red/blob/master/red.js#L163-L168
- https://github.com/node-red/node-red/blob/master/red.js#L183
- https://github.com/node-red/node-red/blob/master/red.js#L200
- https://github.com/node-red/node-red/blob/master/red.js#L274
- https://github.com/node-red/node-red/blob/master/red.js#L317

Or a look into the most changed file, `view.js`:
https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L78-L79
https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L93
https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L104
https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L122

**Comment by eisbehr-** [comment]
> rather than **highlight the benefits** of using such a tool.

...

> you **don't need to explain the benefits**.

What now?!

------

And I would not confirm, that the rest of the code looks better. Lets make a quick scroll-to the main `red.js` file:
- https://github.com/node-red/node-red/blob/master/red.js#L18-L19
- https://github.com/node-red/node-red/blob/master/red.js#L22-L23
- https://github.com/node-red/node-red/blob/master/red.js#L116-L117
- https://github.com/node-red/node-red/blob/master/red.js#L163-L168
- https://github.com/node-red/node-red/blob/master/red.js#L183
- https://github.com/node-red/node-red/blob/master/red.js#L200
- https://github.com/node-red/node-red/blob/master/red.js#L274
- https://github.com/node-red/node-red/blob/master/red.js#L317

Or a look into the most changed file, `view.js`:
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L78-L79
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L93
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L104
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L122
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L148
- https://github.com/node-red/node-red/blob/master/editor/js/ui/view.js#L215-L216
- and so on ...

I will be silent now on this, because you don't want this here. It was just an hind/idea. Feel free to close this issue here and mybe discuss this on your own. I'm fine with it ...

**Comment by knolleary** [comment]
I apologise - I see my comments appear to be conflicting, that was not my intent.

We are familiar with the benefits of linting which is what I meant by not needing you to tell us about them. My feedback in the original response was that your original issue came across as quite aggressive and could have taken a more constructive approach.

Pointing out lots of minor examples in the code where we haven't been entirely consistent does not help with the discussion.

Thank you for bring this topic up. As I've said, it was one we are aware of and had chosen not to do anything about at this time whilst it has been such a small group of contributors to the core. As we want to encourage more contributors to the code base, I recognise having linting in place will help improve and maintain a consistent style.

**Comment by knolleary** [comment]
https://github.com/node-red/node-red/issues/5794
